arch/x86/crc32_fold_pclmulqdq.c

```diff
@@ -24,12 +25,18 @@
 #include <smmintrin.h> // _mm_extract_epi32
 
 #include "x86_features.h"
+#include "cpu_features.h"
 
 #include "../../crc32_fold.h"
+#include "../../crc32_p.h"
+#include <assert.h>
 
 #ifdef X86_VPCLMULQDQ_CRC
 extern size_t fold_16_vpclmulqdq(__m128i *xmm_crc0, __m128i *xmm_crc1,
     __m128i *xmm_crc2, __m128i *xmm_crc3, uint8_t *dst, const uint8_t *src, size_t len);
+extern size_t fold_16_vpclmulqdq_nocp(__m128i *xmm_crc0, __m128i *xmm_crc1,
+    __m128i *xmm_crc2, __m128i *xmm_crc3, const uint8_t *src, size_t len, __m128i init_crc,
```

Comment:
I thought you were going to just have one `fold_16` function with `if (dst == NULL)` checks. What happened to that?

Reply:
I wanted to avoid having yet another set of branches in the code to jump over the stores.

Reply:
So it's been weighing on my mind a bit - the templated form of this probably wouldn't be too ugly, though the signatures would probably be different, slightly.  I'll see about opening up another PR once this is merged, I should be able to unify this a bit.

And, if this method works out, I'd like to try a few other things:

- There's an obvious place to do exactly the same thing for the adler checksums in inflate_p.h
- aarch64/neon has a vectorized carryless multiplication instruction that can be used as a wider CRC calculation as well.  We could very easily write a version that copies and computes the checksum at the same time for aarch64/neon, where the copy is also similarly nearly free